### PR TITLE
Use CRI interface instead of docker calls in network diagnostics

### PR DESCRIPTION
**Pull request by pravisankar**
Trello card: https://trello.com/c/YlLoBpg5/16-5-make-network-diagnostics-work-with-cri-o-runtime

### CODE REVIEW

**Comment by pravisankar** [comment]
@openshift/sig-network-edge @knobunc @sosiouxme  PTAL, ready for review/merge

**Comment by pravisankar** [comment]
@smarterclayton Bumped oc binary limit size check from 115 to 116M, can you please approve?

File: pkg/oc/admin/diagnostics/diagnostics/cluster/network/in_pod/pod.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	kapi "k8s.io/kubernetes/pkg/apis/core"
 	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kcontainer "k8s.io/kubernetes/pkg/kubelet/container"
```
Comment (smarterclayton) [inline pkg/oc/admin/diagnostics/diagnostics/cluster/network/in_pod/pod.go]: No, you are not allowed to add this package set to `oc`.  Diagnostics may not depend on any apiserver, controller manager, or kubelet code.

/hold

**Comment by smarterclayton** [comment]
Unfortunately you may not take references to kubelet code (at all) from oc.

**Comment by knobunc** [comment]
@smarterclayton is there a suggested path forwards?

**Comment by smarterclayton** [comment]
Diagnostics is limited to what you can via the api and via a container.

I would not recommend growing use of diagnostics if you can

1. Add a metric
2. Have the network plugin expose a protected debug endpoint that reports
this data

On Jul 13, 2018, at 1:54 PM, Ben Bennett <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> is there a suggested
path forwards?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/19216#issuecomment-404906858>,
or mute
the thread

**Comment by knobunc** [comment]
@smarterclayton we are not trying to grow diags here, just keep parity between docker and cri-o.

**Comment by pravisankar** [comment]
Removed kubelet package dependency from network diagnostics cli command.
@knobunc @ironcladlou @sosiouxme PTAL

File: pkg/oc/cli/admin/diagnostics/diagnostics/cluster/network/in_pod/util/runtime.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
+package util
```
Comment (ironcladlou) [inline pkg/oc/cli/admin/diagnostics/diagnostics/cluster/network/in_pod/util/runtime.go]: I think I would have gone with an interface/implementations approach here, but I also think we should be using a shared library for container abstractions (which @smarterclayton already nixed in the form of a kubelet dependency)... so I'm not going to argue much with the implementation design. @smarterclayton, is this better than allowing the kubelet dep? Is there some other runtime interface outside the kubelet we could be using?

**Comment by ironcladlou** [comment]
lgtm (with some nits about the `Runtime` design which I'm not going to argue about), will let @knobunc give the final signoff

**Comment by pravisankar** [comment]
To avoid importing kubelet package, updated commit is using crictl and docker CLI commands. If we have used CRI api interface, we could have supported any runtime without any changes. Current implementation works for docker and cri-o runtimes and support for any other runtime needs additional changes. I think that's the trade off for keeping oc binary small with this implementation.

@knobunc if you are okay with this limitation, can you lgtm/approve the changes?

**Comment by smarterclayton** [comment]
Shell callouts are fine.

**Comment by ironcladlou** [comment]
Spoke with @smarterclayton and he's good with the new approach, let's get this moving

**Pull request merged**